### DISCUSSION TITLE
Not able to build after updating wagmi

### DISCUSSION THREAD

**Discussion by dcalap** [wevm/wagmi#2248] [Q&A]
After updating my wagmi version (from 0.12.6 to 0.12.11 and as well with 0.12.12) in my vite project, I'm getting this error when building:

`[vite]: Rollup failed to resolve import "@safe-globalThis/safe-apps-provider" from "node_modules/@wagmi/connectors/dist/safe.js".
This is most likely unintended because it can break your application at runtime.
If you do want to externalize this module explicitly add it to
`build.rollupOptions.external`
error during build:
Error: [vite]: Rollup failed to resolve import "@safe-globalThis/safe-apps-provider" from "node_modules/@wagmi/connectors/dist/safe.js".
This is most likely unintended because it can break your application at runtime.
If you do want to externalize this module explicitly add it to
`build.rollupOptions.external`
`

Any idea why and how to solve it?

Thanks a lot!

**Comment by tmm** [comment]
Can you post a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example)? Hard to debug without seeing any of the code.

**Comment by A-M-H-F** [comment]
Add this to vite.config.ts *FOR BUILD ONLY*

replace:
`
  define: {
    global: 'globalThis',
  }
`


with:
```javascript
define: {
    global: (() => {
      let globalVariable = 'globalThis';
      try {
        // Try to import @safe-global/safe-apps-provider
        require.resolve('@safe-global/safe-apps-provider');
        // Try to import @safe-global/safe-apps-sdk
        require.resolve('@safe-global/safe-apps-sdk');
        // If both modules are found, return the custom global variable
        globalVariable = 'global';
       } catch (e) {
        // If either module is not found, fallback to globalThis
        globalVariable = 'globalThis';
       }
      return globalVariable;
     })()
   }
```

**Reply by Willzxxz** [reply]
thanks

**Reply by Willzxxz** [reply]
```
define: {
    global: (() => {
      let globalVariable = "globalThis";
      try {
        // Try to import @safe-global/safe-apps-provider
        require.resolve("@safe-global/safe-apps-provider");
        // Try to import @safe-global/safe-apps-sdk
        require.resolve("@safe-global/safe-apps-sdk");
        // If both modules are found, return the custom global variable
        globalVariable = "global";
      } catch (e) {
        // If either module is not found, fallback to globalThis
        globalVariable = "globalThis";
      }
      return globalVariable;
    })(),
  }
```

**Reply by davidcalap** [reply]
Thanks a lot, I just want to know what you mean by `FOR BUILD ONLY`. I'm uploading my app to Heroku, so, should I commit this change? Thanks

**Reply by A-M-H-F** [reply]
> Thanks a lot, I just want to know what you mean by `FOR BUILD ONLY`. I'm uploading my app to Heroku, so, should I commit this change? Thanks

Only use this when you are building a production version, when running as dev use globalThis as value, if you need to solve this completely, try to install everything from scratch using wagmi create,this error is happening when you add wagmi after creating (vite, react...).

**Reply by kingcozz** [reply]
same issue. nothing works and can no longer deploy to vercel

**Reply by alexboisseau** [reply]
to have only one config file which handle the "FOR BUILD ONLY", we can use the command params like that : 
```typescript
export default defineConfig(({ command }) => ({
  define: {
    global: (() => {
      if (command !== "build") return "globalThis";
      let globalVariable = "globalThis";
      try {
        // Try to import @safe-global/safe-apps-provider
        require.resolve("@safe-global/safe-apps-provider");
        // Try to import @safe-global/safe-apps-sdk
        require.resolve("@safe-global/safe-apps-sdk");
        // If both modules are found, return the custom global variable
        globalVariable = "global";
      } catch (e) {
        // If either module is not found, fallback to globalThis
        globalVariable = "globalThis";
      }
      return globalVariable;
    })(),
  },
  ...
}));
```

**Comment by oveddan** [comment]
same issue here, using wagmi 1.0.6; i just used wagmi cli to create a new project, using the vite + rainbowkit template; then attempted to deploy it right away to github pages

**Reply by KneePham** [reply]
I'm having the same issue on a none template install. I've tested with a clean one and it seems to work with vite+rainbow connect. You may still need to replace define: `{ global: 'globalThis', }` with what willzxxz mention above.

**Comment by Kayeddy** [comment]
Hello, for all of you who are still facing this error, I found a solution on StackOverflow that helped me fix the issue. Here is the link: https://stackoverflow.com/questions/76208402/vite-rollup-failed-to-resolve-import-src-main-tsx

To summarize, there are two things y'all should do: 

1. Move your index.html to the root folder of your project.
2. Modify the Vite config file as follows: 

```
import { defineConfig } from 'vite'
import react from '@vitejs/plugin-react'

// https://vitejs.dev/config/
export default defineConfig({
  plugins: [react()],
  build:{
    outDir: "build"
  }
})
```
Note that the outDir property can be "build" or "dist" depending on your setup.  After implementing the previous steps try to build your project again using `npm run build`, `vite build` or `npx vite build`

---------------------------------------------------------------------------------

**Edit:** 

I forgot to mention, the Vite configuration I provided should only be used for production. When it comes to development and debugging you should use the config provided by Vite's website. 

In my case, for example, I have my Vite config file like this: 

```
import { defineConfig } from "vite";
import react from "@vitejs/plugin-react";

// Development config
// https://vitejs.dev/config/
export default defineConfig({
  plugins: [react()],
  define: {
    global: "globalThis",
    "process.env": {},
  },
});

// Production config
/*
export default defineConfig({
  plugins: [react()],
  build: {
    outDir: "dist",
  },
});
*/
```

Whenever I build for production I just comment the 'Development config' section and uncomment the 'Production config' section.

I do hope this helps. Kind regards!

**Reply by gosuto-inzasheru** [reply]
so true solution is to create a second config, ie `vite.config.build.ts`, scrap the `global: "globalThis"` from it and then in your `package.json` redefine your alias so that you build using this second config file: `vite build --config vite.config.build.ts` :)

**Comment by chknlittle** [comment]
> so true solution is to create a second config, ie `vite.config.build.ts`, scrap the `global: "globalThis"` from it and then in your `package.json` redefine your alias so that you build using this second config file: `vite build --config vite.config.build.ts` :)

this is the right answer than you @gosuto-inzasheru
